presto-parser/src/main/java/io/prestosql/sql/parser/AstBuilder.java

```diff
@@ -580,6 +580,13 @@ public Node visitQueryNoWith(SqlBaseParser.QueryNoWithContext context)
             orderBy = Optional.of(new OrderBy(getLocation(context.ORDER()), visit(context.sortItem(), SortItem.class)));
         }
 
+        Optional<String> limit = getTextIfPresent(context.limit);
+        if (context.FETCH() != null) {
+            limit = Optional.of(limit.orElse("1"));  // default FETCH row count is 1
+            long limitValue = Long.parseLong(limit.get());
+            check(limitValue != 0, "Invalid row count in FETCH FIRST clause", context.limit);
```

Comment:
This is a semantic check that should be deferred until analysis.

Reply:
The semantic difference between LIMIT and FETCH is whether they allow row count equal to 0.

In my approach, in the course of processing the query, FETCH and LIMIT are distinguishable only as far as `AstBuilder`. That is why I put the semantic check (row count > 0) there. 

To avoid that, LIMIT and FETCH could be carried distinctly until analysis which would enable FETCH-specific check there, and glued together afterwards. I consider this over-complicating, as they are essentially the same. 

Or I could define a new token in grammar like `POSITIVE_INTEGER` and let parser do the check.

Reply:
I guess that works until we add percentage and WITH TIES.

Also, technically, the number can be arbitrarily large. The spec requires it to be "an exact numeric type with scale 0" (for rows -- for percent it can be any numeric type). So `Long.parseLong` might not be sufficient. Incidentally, specifying a larger value will fail with a `NumberFormatException`. Validating this in the analyzer allows us to provide better error messages and error codes.

Another issue is that parsing/formatting doesn't roundtrip. If you have a query with "FETCH FIRST x ROWS" it will be printed as "LIMIT x" by the SqlFormatter.

Reply:
OK, I see that there is more to analyse than just row count greater than zero or not. So the right approach is to keep LIMIT and FETCH distinct until analysis.
I think I'll keep them separate until `QueryPlanner` and there produce a `LimitNode` for a query with `fetch`.